xpdacq/plans.py

```diff
@@ -0,0 +1,126 @@
+    percentile : float
+        Percentile to use to measure the max intensity (use percentile because
+        of bad pixels)
+    lower_threshold : float
+        Threshold below which data is unreliable for measuring sample
+        scattering power
+    desired_exposure : float
+        The desired exposure. The exposure must be within the capabilities of
+        the detector. This will also be used to sample the sample's scattering
+        power.
+
+    Returns
+    -------
+    plan : generator
+        The plan
+
+    Notes
+    -----
+    This requires accurate attenuation factors for the filters.
+
+    """
+    # List of filter configurations in order of filter power
+    # TODO: Need to set this since it could be non-unique
+    filter_configurations = sorted(
+        [np.asarray(x) for x in product([0, 1], repeat=4)],
+        key=lambda x: np.prod((x * fb_attenuation)[np.nonzero(x)]),
+        reverse=True,
+    )
+    attenuations = 1 / np.asarray(
+        [
+            np.prod((x * fb_attenuation)[np.nonzero(x)])
+            for x in filter_configurations
+        ]
+    )
+
+    @bpp.stage_decorator([xpd_configuration["area_det"]])
+    @bpp.run_decorator()
+    def inner():
+        yield from bps.abs_set(
+            xpd_configuration["area_det"].cam.acquire_time, desired_exposure
+        )
+        for fc in filter_configurations:
+            att = 1 / np.prod((fc * fb_attenuation)[np.nonzero(fc)])
+            # set the filter banks
+            yield from (
+                bps.abs_set(f, io, wait=True)
+                for f, io in zip(xpd_configuration["filter_bank"], fc)
+            )
+            data = yield from dark_light()
+
+            # calculate the value of the top .01 percentile pixel
+            v = np.percentile(data, percentile)
+
+            # The value is too low to evaluate, bump to next configuration
+            if v < lower_threshold:
+                continue
+            else:
+                break
+        # compute the sample scattering power
+        sp = v / (att * desired_exposure)
+        # compute filter config index to use
+        fci = np.where(
+            (sp * attenuations * desired_exposure) < upper_threshold
+        )[0][-1]
+        # set the bank
+        yield from (
+            bps.abs_set(f, io, wait=True)
+            for f, io in zip(
+                xpd_configuration["filter_bank"], filter_configurations[fci]
+            )
+        )
+
+        # compute the exposure
+        computed_exposure = upper_threshold / (sp * attenuations[fci])
+        print("Extrapolated exposure: {}".format(computed_exposure))
+        exposure = min(5, computed_exposure)
+        print("Using exposure: {}".format(computed_exposure))
+        expected_counts = sp * attenuations[fci] * exposure
+        print("Expected counts: {}".format(expected_counts))
+        yield from bps.abs_set(
+            xpd_configuration["area_det"].cam.acquire_time, exposure
+        )
+
```

Comment:
Run another round to double check.

Reply:
reviewed on train together with CJ.  Looks promising, needs testing at the beamline.

We need an implementation/release plan when it is tested....I would like it to be largely hidden functionality and not change (enormously) the UI but this needs to be discussed. (we will need an override kwarg)

Reply:
Also waiting a lightweight databroker for loading the light/dark data live.

Reply:
xref: https://github.com/NSLS-II/bluesky/issues/1079